bundles/cif-virtual-catalog/src/main/java/com/adobe/cq/commerce/virtual/catalog/data/impl/CatalogDataResourceProviderManagerImpl.java

```diff
@@ -478,11 +489,15 @@ void unbindFactory(CatalogDataResourceProviderFactory factory, Map<String, Strin
 
     @Override
     public List<Resource> getDataRoots() {
-        return dataRoots != null ? dataRoots : findDataRoots(resolver);
+        List<Resource> roots = new ArrayList<>();
+        synchronized (dataRoots) {
```

Comment:
When using a CopyOnWirteArrayList you don't need synchronisation here. I assume this is more often called than  registerDataRoot(s) and so synchronisation would have a negative impact.